File: sdk/typescript/src/transport/http.ts
import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import { createParser } from 'eventsource-parser';
import { IntelliRouterConfig, RequestOptions, HttpMethod } from '../types';
import { Transport } from './base';
import {
    ApiError,
    AuthenticationError,
    AuthorizationError,
    NotFoundError,
    RateLimitError,
    ServerError,
    TimeoutError,
    ValidationError,
} from '../errors';

/**
 * HTTP transport implementation
 */
export class HttpTransport implements Transport {
    private readonly client: AxiosInstance;
    private readonly config: IntelliRouterConfig;

    /**
     * Create a new HTTP transport
     * @param config Client configuration
     */
    constructor(config: IntelliRouterConfig) {
        this.config = config;

        this.client = axios.create({
            baseURL: config.baseUrl || 'http://localhost:8000',
            timeout: config.timeout || 30000,
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
                ...(config.apiKey ? { 'Authorization': `Bearer ${config.apiKey}` } : {}),
                ...(config.defaultHeaders || {}),
            },
        });

        // Add request interceptor for retries
        this.client.interceptors.response.use(
            (response) => response,
            async (error) => {
                if (!error.config || !this.shouldRetry(error)) {
                    return Promise.reject(error);
                }

                const retryCount = error.config.retryCount || 0;
                const maxRetries = this.config.maxRetries || 3;

                if (retryCount >= maxRetries) {
                    return Promise.reject(error);
                }

                error.config.retryCount = retryCount + 1;

                // Exponential backoff
                const delay = Math.pow(2, retryCount) * 1000;
                await new Promise((resolve) => setTimeout(resolve, delay));

                return this.client(error.config);
            }
        );
    }

    /**
     * Determine if a request should be retried
     * @param error Axios error
     * @returns Whether the request should be retried
     */
    private shouldRetry(error: any): boolean {
        // Don't retry if we don't have a response
        if (!error.response) {
            return false;
        }

        // Don't retry client errors except for 429 (rate limit)
        if (error.response.status >= 400 && error.response.status < 500 && error.response.status !== 429) {
            return false;
        }

        return true;
    }

    /**
     * Handle an error response
     * @param error Axios error
     * @throws ApiError or a subclass
     */
    private handleError(error: any): never {
        if (error.response) {
            const { status, data } = error.response;
            const message = data?.error?.message || data?.message || error.message || 'Unknown error';

            switch (status) {
                case 400:
                    throw new ValidationError(message, { details: data });
                case 401:
                    throw new AuthenticationError(message, { details: data });
                case 403:
                    throw new AuthorizationError(message, { details: data });
                case 404:
                    throw new NotFoundError(message, { details: data });
                case 429:
                    throw new RateLimitError(message, { details: data });
                default:
                    if (status >= 500) {
                        throw new ServerError(message, { status, details: data });
                    }
                    throw new ApiError(message, { code: 'api_error', status, details: data });
            }
        } else if (error.code === 'ECONNABORTED') {
            throw new TimeoutError('Request timed out');
        } else {
            throw new ApiError(error.message || 'Unknown error', { code: 'network_error', status: 0 });
        }
    }

    /**
     * Send a request
     * @param options Request options
     * @returns Promise resolving to the response data
     */
    public async request<T>(options: RequestOptions): Promise<T> {
        try {
            const config: AxiosRequestConfig = {
                method: options.method,
                url: options.path,
                params: options.params,
                data: options.body,
                headers: options.headers,
            };

            const response = await this.client.request<T>(config);
            return response.data;
        } catch (error) {
            this.handleError(error);
        }
    }

    /**
     * Send a request with streaming response
     * @param options Request options
     * @returns Promise resolving to an async iterable of response chunks
     */
    public async requestStream(options: RequestOptions): Promise<AsyncIterable<unknown>> {
        try {
            const config: AxiosRequestConfig = {
                method: options.method,
                url: options.path,
                params: options.params,
                data: options.body,
                headers: {
                    ...options.headers,
                    'Accept': 'text/event-stream',
                },
                responseType: 'stream',
            };

            const response = await this.client.request(config);

            return this.parseEventStream(response);
        } catch (error) {
            this.handleError(error);
        }
    }

    /**
     * Parse an event stream response
     * @param response Axios response
     * @returns Async iterable of parsed events
     */
    private async *parseEventStream(response: AxiosResponse): AsyncIterable<unknown> {
        const parser = createParser((event) => {
            if (event.type === 'event' && event.data) {
                if (event.data === '[DONE]') {
                    return;
                }

                try {
                    const parsedData = JSON.parse(event.data);
                    return parsedData;
                } catch (e) {
                    return event.data;
                }
            }
        });

        const stream = response.data;

        for await (const chunk of stream) {
            const str = chunk.toString();
            const parsedEvents = parser.feed(str);

            for (const event of parsedEvents) {
                if (event) {
                    yield event;
                }
            }
        }
    }

    /**
     * Send a GET request
     * @param path URL path
     * @param params Query parameters
     * @param headers Request headers
     * @returns Promise resolving to the response data
     */
    public async get<T>(path: string, params?: Record<string, string>, headers?: Record<string, string>): Promise<T> {
        return this.request<T>({
            method: 'GET',
            path,
            params,
            headers,
        });
    }

    /**
     * Send a POST request
     * @param path URL path
     * @param body Request body
     * @param headers Request headers
     * @returns Promise resolving to the response data
     */
    public async post<T>(path: string, body?: unknown, headers?: Record<string, string>): Promise<T> {
        return this.request<T>({
            method: 'POST',
            path,
            body,
            headers,
        });
    }

    /**
     * Send a PUT request
     * @param path URL path
     * @param body Request body
     * @param headers Request headers
     * @returns Promise resolving to the response data
     */
    public async put<T>(path: string, body?: unknown, headers?: Record<string, string>): Promise<T> {
        return this.request<T>({
            method: 'PUT',
            path,
            body,
            headers,
        });
    }

    /**
     * Send a DELETE request
     * @param path URL path
     * @param headers Request headers
     * @returns Promise resolving to the response data
     */
    public async delete<T>(path: string, headers?: Record<string, string>): Promise<T> {
        return this.request<T>({
            method: 'DELETE',
            path,
            headers,
        });
    }

    /**
     * Send a POST request with streaming response
     * @param path URL path
     * @param body Request body
     * @param headers Request headers
     * @returns Promise resolving to an async iterable of response chunks
     */
    public async postStream(path: string, body?: unknown, headers?: Record<string, string>): Promise<AsyncIterable<unknown>> {
        return this.requestStream({
            method: 'POST',
            path,
            body,
            headers,
            stream: true,
        });
    }
}

File: sdk/python/transport/http.py
from typing import Dict, Any, Optional, Union, AsyncIterator, Iterator
import requests
import aiohttp
import json
import sseclient
import asyncio
from ..config import Configuration
from ..exceptions import APIError, AuthenticationError, RateLimitError, ServerError
from .base import Transport

class HTTPTransport(Transport):
    """
    HTTP transport layer for making requests to the IntelliRouter API.
    
    Args:
        config: Configuration object.
    """
    
    def __init__(self, config: Configuration):
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json",
        })
    
    def request(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
        stream: bool = False,
    ) -> Dict[str, Any]:
        """
        Make a synchronous request to the IntelliRouter API.
        
        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path.
            params: Query parameters.
            data: Request body.
            stream: Whether to stream the response.
        
        Returns:
            Response data.
        
        Raises:
            APIError: If the API returns an error.
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
        """
        url = f"{self.config.base_url}{path}"
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                params=params,
                json=data,
                stream=stream,
                timeout=self.config.timeout,
            )
            
            if stream:
                return response
            
            if response.status_code >= 400:
                self._handle_error_response(response)
            
            return response.json()
        except requests.exceptions.RequestException as e:
            raise APIError(f"Request failed: {str(e)}")
    
    async def arequest(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
        stream: bool = False,
    ) -> Dict[str, Any]:
        """
        Make an asynchronous request to the IntelliRouter API.
        
        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path.
            params: Query parameters.
            data: Request body.
            stream: Whether to stream the response.
        
        Returns:
            Response data.
        
        Raises:
            APIError: If the API returns an error.
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
        """
        url = f"{self.config.base_url}{path}"
        
        async with aiohttp.ClientSession() as session:
            headers = {
                "Authorization": f"Bearer {self.config.api_key}",
                "Content-Type": "application/json",
            }
            
            try:
                async with session.request(
                    method=method,
                    url=url,
                    params=params,
                    json=data,
                    headers=headers,
                    timeout=self.config.timeout,
                ) as response:
                    if stream:
                        return response
                    
                    if response.status >= 400:
                        await self._ahandle_error_response(response)
                    
                    return await response.json()
            except aiohttp.ClientError as e:
                raise APIError(f"Request failed: {str(e)}")
    
    def stream(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
    ) -> Iterator[Dict[str, Any]]:
        """
        Make a streaming synchronous request to the IntelliRouter API.
        
        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path.
            params: Query parameters.
            data: Request body.
        
        Returns:
            Iterator of response chunks.
        
        Raises:
            APIError: If the API returns an error.
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
        """
        response = self.request(method, path, params, data, stream=True)
        
        if response.status_code >= 400:
            self._handle_error_response(response)
        
        client = sseclient.SSEClient(response)
        
        for event in client.events():
            if event.data == "[DONE]":
                break
            
            try:
                yield json.loads(event.data)
            except json.JSONDecodeError:
                raise APIError(f"Invalid JSON in stream: {event.data}")
    
    async def astream(
        self,
        method: str,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        data: Optional[Dict[str, Any]] = None,
    ) -> AsyncIterator[Dict[str, Any]]:
        """
        Make a streaming asynchronous request to the IntelliRouter API.
        
        Args:
            method: HTTP method (GET, POST, etc.).
            path: API path.
            params: Query parameters.
            data: Request body.
        
        Returns:
            Async iterator of response chunks.
        
        Raises:
            APIError: If the API returns an error.
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
        """
        response = await self.arequest(method, path, params, data, stream=True)
        
        if response.status >= 400:
            await self._ahandle_error_response(response)
        
        async for line in response.content:
            line = line.decode("utf-8").strip()
            
            if not line:
                continue
            
            if line.startswith("data:"):
                data = line[5:].strip()
                
                if data == "[DONE]":
                    break
                
                try:
                    yield json.loads(data)
                except json.JSONDecodeError:
                    raise APIError(f"Invalid JSON in stream: {data}")
    
    def _handle_error_response(self, response: requests.Response) -> None:
        """
        Handle an error response from the API.
        
        Args:
            response: Response object.
        
        Raises:
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
            APIError: For other API errors.
        """
        try:
            error_data = response.json()
            error_message = error_data.get("error", {}).get("message", "Unknown error")
        except (json.JSONDecodeError, KeyError):
            error_message = response.text or "Unknown error"
        
        if response.status_code == 401:
            raise AuthenticationError(f"Authentication failed: {error_message}")
        elif response.status_code == 429:
            raise RateLimitError(f"Rate limit exceeded: {error_message}")
        elif response.status_code >= 500:
            raise ServerError(f"Server error: {error_message}")
        else:
            raise APIError(f"API error: {error_message}", response.status_code)
    
    async def _ahandle_error_response(self, response: aiohttp.ClientResponse) -> None:
        """
        Handle an error response from the API asynchronously.
        
        Args:
            response: Response object.
        
        Raises:
            AuthenticationError: If authentication fails.
            RateLimitError: If the rate limit is exceeded.
            ServerError: If the server returns an error.
            APIError: For other API errors.
        """
        try:
            error_data = await response.json()
            error_message = error_data.get("error", {}).get("message", "Unknown error")
        except (json.JSONDecodeError, KeyError):
            error_message = await response.text() or "Unknown error"
        
        if response.status == 401:
            raise AuthenticationError(f"Authentication failed: {error_message}")
        elif response.status == 429:
            raise RateLimitError(f"Rate limit exceeded: {error_message}")
        elif response.status >= 500:
            raise ServerError(f"Server error: {error_message}")
        else:
            raise APIError(f"API error: {error_message}", response.status)